datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/mirrormaker/KafkaMirrorMakerConnector.java

```diff
@@ -109,4 +155,189 @@ public void postDatastreamInitialize(Datastream datastream, List<Datastream> all
+    private final Pattern _topicPattern;
+
+    // The partitions covered by this datastresm group, fetched from Kafka
+    private volatile List<String> _subscribedPartitions = Collections.<String>emptyList();
+
+    // indicate if this thread has already fetch the partitions info from Kafka
+    private volatile boolean _initialized;
+
+
+    private PartitionDiscoveryThread(DatastreamGroup datastreamGroup) {
+      _datastreamGroup = datastreamGroup;
+      //Compile topic pattern so that it contains the topic regex from source KafkaConnectionString
+      //Example: source string:  kafka://HOST:9092/^test.*$, topic pattern: ^test.*$
+      _topicPattern = Pattern.compile(
+          KafkaConnectionString.valueOf(_datastreamGroup.getDatastreams().get(0).getSource().getConnectionString()).getTopicName());
+      _initialized = false;
+    }
+
+
+    private List<String> getPartitionsInfo(Consumer<?, ?> consumer) {
+      Map<String, List<PartitionInfo>> sourceTopics = consumer.listTopics();
+      List<TopicPartition> topicPartitions = sourceTopics.keySet().stream()
+          .filter(t1 -> _topicPattern.matcher(t1).matches())
+          .flatMap(t2 ->
+              sourceTopics.get(t2).stream().map(partitionInfo ->
+                  new TopicPartition(partitionInfo.topic(), partitionInfo.partition()))).collect(Collectors.toList());
+
+      return topicPartitions.stream().map(TopicPartition::toString).sorted().collect(Collectors.toList());
+    }
+
+    private Consumer<?, ?> createConsumer(Properties consumerProps, String bootstrapServers, String groupId) {
+      Properties properties = new Properties();
+      properties.putAll(consumerProps);
+      properties.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
+      properties.put(ConsumerConfig.GROUP_ID_CONFIG, groupId);
+      properties.putIfAbsent(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,
+          ByteArrayDeserializer.class.getCanonicalName());
+      properties.putIfAbsent(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
+          ByteArrayDeserializer.class.getCanonicalName());
+      return _listenerConsumerFactory.createConsumer(properties);
+    }
+
+    @Override
+    public void run() {
+      Datastream datastream = _datastreamGroup.getDatastreams().get(0);
+      String bootstrapValue = KafkaConnectionString.valueOf(datastream.getSource().getConnectionString())
+          .getBrokers().stream()
+          .map(KafkaBrokerAddress::toString)
+          .collect(Collectors.joining(KafkaConnectionString.BROKER_LIST_DELIMITER));
+      Consumer<?, ?> consumer = createConsumer(_consumerProperties, bootstrapValue,
+          _groupIdConstructor.constructGroupId(datastream) + DEST_CONSUMER_GROUP_ID_SUFFIX);
+
+      LOG.info("Fetch thread for {} started", _datastreamGroup.getName());
+      while (!isInterrupted() && !_shutdown) {
+        try {
+          // If partition is changed
+          List<String> newPartitionInfo = getPartitionsInfo(consumer);
```

Comment:
unlike consumer.poll(), consumer.listTopics() doesn't involve rebalance it has much less chance to stuck. By default, Kafka supplies a default api timeout which is 60s. So in the worst case, this will throw timeout after 60 seconds and get retried again in the next iteration.